tools-containers/kdc-api-server/server/server.go

```diff
@@ -0,0 +1,593 @@
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"strings"
+)
+
+type KDCAPIServer struct {
+	kadmin   *KAdminClient
+	endpoint string
+}
+
+type KDCRequestAddPrincipal struct {
+	PrincipalsRaw string       `json:"principals_raw"`
+	Principals    []KPrincipal `json:"principals"`
+	Secret        string       `json:"secret"`
+	Binary        *bool        `json:"binary"`
+}
+
+type KDCRequestListPrincipals struct {
+	Filter string `json:"filter"`
+	Secret string `json:"secret,omitempty`
+	Binary *bool  `json:"binary"`
+}
+
+type KDCCheckStatus struct {
+	Pass   bool   `json:"pass"`
+	Reason string `json:"reason,omitempty"`
+}
+
+type KDCResponse struct {
+	Status     string          `json:"status"`
+	Error      string          `json:"error,omitempty"`
+	Principals []KPrincipal    `json:"principals,omitempty"`
+	Check      *KDCCheckStatus `json:"check,omitempty"`
+}
+
+func createKDCAPIServer(kadmin *KAdminClient, port string, host string) *KDCAPIServer {
+	inst := &KDCAPIServer{
+		kadmin:   kadmin,
+		endpoint: fmt.Sprintf("%s:%s", host, port),
+	}
+
+	// Register the static API server
+	fs := http.FileServer(http.Dir("static/"))
+	http.Handle("/static/", http.StripPrefix("/static/", fs))
+	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintf(w, "Welcome to KDC Admin")
+	})
+
+	// Register the app routes
+	http.HandleFunc("/api/principals", inst.handlePrincipals)
+	http.HandleFunc("/api/check", inst.handleCheckPrincipals)
+	return inst
+}
+
+func (s *KDCAPIServer) Start() {
+	log.Printf("Listening on %s\n", s.endpoint)
+	http.ListenAndServe(s.endpoint, nil)
+}
+
+/**
+ * Respond with an error, formatting it according to the request type
+ */
+func (s *KDCAPIServer) replyReject(rw http.ResponseWriter, req *http.Request, format string, args ...interface{}) {
+	contentType := req.Header.Get("Content-Type")
+	if contentType == "application/json" {
+		resp := KDCResponse{
+			"error",
+			fmt.Sprintf(format, args...),
+			nil,
+			nil,
+		}
+
+		// Marshal the data
+		js, err := json.Marshal(resp)
+		if err != nil {
+			http.Error(rw, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		// Respond with JSON-encoded text
+		rw.Header().Set("Content-Type", "application/json")
+		rw.Write(js)
+	} else {
+		// Otherwise respond with raw text
+		rw.Header().Set("Content-Type", "text/plain")
+		fmt.Fprintf(rw, "errr: "+format, args...)
+	}
+}
+
+func (s *KDCAPIServer) replySuccess(rw http.ResponseWriter, req *http.Request, data interface{}) {
+	contentType := req.Header.Get("Content-Type")
+	if contentType == "application/json" {
+		resp := KDCResponse{
+			"ok", "", nil, nil,
+		}
+
+		if data != nil {
+			switch x := data.(type) {
+			case []KPrincipal:
+				resp.Principals = x
+			case *KDCCheckStatus:
+				resp.Check = x
+			}
+		}
+
+		// Marshal the data
+		js, err := json.Marshal(resp)
+		if err != nil {
+			http.Error(rw, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		// Respond with JSON-encoded text
```

Comment:
Please remove the comment which states the obvious.

Reply:
Not really. This is a clarification that if the body was "JSON" it should reply with "JSON", otherwise it should reply with plain-text. Restructured this a bit.